docs/port-karpenter-on-our-cloud.md

```diff
@@ -0,0 +1,90 @@
+# Porting Karpenter to your cloud provider
+
+## Pre-requisites
+
+Karpenter requires your to implement an interface to use your own cloud provider.
+The interface is defined in the `pkg/cloudprovider/types.go` file.
+
+```go
+// CloudProvider interface is implemented by cloud providers to support provisioning.
+type CloudProvider interface {
+	// Create launches a NodeClaim with the given resource requests and requirements and returns a hydrated
+	// NodeClaim back with resolved NodeClaim labels for the launched NodeClaim
+	Create(context.Context, *v1beta1.NodeClaim) (*v1beta1.NodeClaim, error)
+	// Delete removes a NodeClaim from the cloudprovider by its provider id
+	Delete(context.Context, *v1beta1.NodeClaim) error
+	// Get retrieves a NodeClaim from the cloudprovider by its provider id
+	Get(context.Context, string) (*v1beta1.NodeClaim, error)
+	// List retrieves all NodeClaims from the cloudprovider
+	List(context.Context) ([]*v1beta1.NodeClaim, error)
+	// GetInstanceTypes returns instance types supported by the cloudprovider.
+	// Availability of types or zone may vary by nodepool or over time.  Regardless of
+	// availability, the GetInstanceTypes method should always return all instance types,
+	// even those with no offerings available.
+	GetInstanceTypes(context.Context, *v1beta1.NodePool) ([]*InstanceType, error)
+	// IsDrifted returns whether a NodeClaim has drifted from the provisioning requirements
+	// it is tied to.
+	IsDrifted(context.Context, *v1beta1.NodeClaim) (DriftReason, error)
+	// Name returns the CloudProvider implementation name.
+	Name() string
+}
+```
+
```

Comment:
Recommended section: How does karpenter core interact with the cloudprovider interface? 

This is very important to understand the patterns the deletes and creates follow. If I create a node on aks, we have to create 4 resources for that node to bootstrap. 


1. network interface 
2. disk 
3. virtual machine 
4. virtual machine extension 


We see that for the nic and disk we will not have problems on create, but for virtual machine create, we may hit some error. 

Understanding that the launch controller will then retry might be useful in understanding how people should handle resource cleanup.